src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,214 @@
+package duke;
+
+
+import duke.exception.*;
+import duke.task.TaskType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Represents the list of tasks together with a counter
+ * to keep track of pending tasks
+ */
+public class TaskList implements Cloneable {
+    private List<Task> tasks;
+    private int numOfPendingTasks;
+    private static final Ui UI = new Ui();
+
+    TaskList() {
+        this.tasks = new ArrayList<>();
+        this.numOfPendingTasks = 0;
+    }
+
+    TaskList(ArrayList<Task> tasks) {
+        this.tasks = tasks;
+        this.numOfPendingTasks = 0;
+        for (Task task : tasks) {
+            if (!task.isDone) {
+                incrementPendingTasks();
+            }
+        }
+    }
+
+    /**
+     * Extracts the list of tasks of a TaskList.
+     *
+     * @return a list of tasks contained in a TaskList object.
+     */
+    public List<Task> getTasks() {
+        return tasks;
+    }
+
+    /**
+     * Checks if a call to task index in the CLI is valid.
+     *
+     * @param target Integer number
+     * @return true if target is usable as an array access
+     */
+    public boolean isWithinValidRange(int target) {
+        return target >= 0 && target < tasks.size();
+    }
+
+    /**
+     * Adds one count to pending task.
+     */
+    public void incrementPendingTasks() {
+        this.numOfPendingTasks++;
+    }
+
+    /**
+     * Subtracts one count to pending task.
+     */
+    public void decrementPendingTasks() {
+        this.numOfPendingTasks--;
+    }
+
+    public int getNumOfPendingTasks() {
```

Comment:
Although the method's function is obvious, it might still be good to add a JavaDoc comment as the method is public.

Reply:
Got it, thanks!